src/SharedPipelineComponents/Datalock/SFA.DAS.CollectionEarnings.DataLock/Services/DatalockValidationService.cs

```diff
@@ -0,0 +1,295 @@
+        {
+            if (period < 6)
+            {
+                return period + 7;
+            }
+
+            return period - 5;
+        }
+
+        private int YearFromPeriod(int period, DateTime episodeStartDate)
+        {
+            var month = MonthFromPeriod(period);
+            var startOfAcademicYear = StartOfAcademicYearFromEpisodeStartDate(episodeStartDate);
+            if (month < 8)
+            {
+                return startOfAcademicYear.Year + 1;
+            }
+            return startOfAcademicYear.Year;
+        }
+
+        private DateTime StartOfAcademicYearFromEpisodeStartDate(DateTime episodeStartDate)
+        {
+            var month = episodeStartDate.Month;
+            if (month < 8)
+            {
+                return new DateTime(episodeStartDate.Year - 1, 8, 1);
+            }
+            return new DateTime(episodeStartDate.Year, 8, 1);
+        }
+    }
+
+    public class DatalockValidationResult
+    {
+        public List<DatalockValidationError> ValidationErrors { get; } = new List<DatalockValidationError>();
+        public List<DatalockValidationErrorByPeriod> ValidationErrorsByPeriod { get; } = new List<DatalockValidationErrorByPeriod>();
+        public List<PriceEpisodePeriodMatchEntity> PriceEpisodePeriodMatches { get; } = new List<PriceEpisodePeriodMatchEntity>();
+        public List<PriceEpisodeMatchEntity> PriceEpisodeMatches { get; } = new List<PriceEpisodeMatchEntity>();
+        public List<DatalockOutputEntity> DatalockOutputEntities { get; } = new List<DatalockOutputEntity>();
+        
+        public void AddDistinctRecords(RawEarning earning, List<string> errors, TransactionTypesFlag paymentType, CommitmentEntity commitment)
+        {
+            var payable = false;
+            if (errors.Count > 0)
+            {
+                foreach (var error in errors)
+                {
+                    if (ValidationErrors.FirstOrDefault(x =>
+                            x.LearnRefNumber == earning.LearnRefNumber &&
+                            x.AimSeqNumber == earning.AimSeqNumber &&
+                            x.PriceEpisodeIdentifier == earning.PriceEpisodeIdentifier &&
+                            x.RuleId == error &&
+                            x.Ukprn == earning.Ukprn) == null &&
+                        PriceEpisodePeriodMatches.FirstOrDefault(x => 
+                            x.AimSeqNumber == earning.AimSeqNumber &&
+                            x.LearnRefNumber == earning.LearnRefNumber &&
+                            x.PriceEpisodeIdentifier == earning.PriceEpisodeIdentifier &&
+                            x.Ukprn == earning.Ukprn && 
+                            x.CommitmentId == commitment.CommitmentId &&
+                            x.Period == earning.Period &&
+                            x.TransactionTypesFlag == paymentType &&
+                            x.Payable &&
+                            x.VersionId == commitment.VersionId) == null)
```

Comment:
This is a very verbose if clause, it looks more like a SQL statement and it would be more readable if it used some methods instead of doing everything at once